better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+program as one for which there exists a valid initial state and
+inputs, such that the program will generate a valid output. As an
+analogy, a broken clock is semi-correct, it is correct twice a
+day. This definition provides a partial ordering of correctness, where
+a program is _more correct_ than another if the initial states under
+which it is correct are a superset of the other.
+
+So while you may not reach a point where all bugs vanish, you will
+have well-justified confidence in the correctness of your own work
+while maintaining high productivity. Also, the discipline we'll
+describe here of **programming by contract** removes uncertainty and
+needless complexity from your code and from the process of coding.
+That's a big part of why we say it's practical.
+
+## Reasoning about correctness
+
+How can you know whether your program is correct?  Validating
+correctness seldom requires an elaborate formal proof—usually the sort
+of everyday thinking that we do while programming suffices.
+
+```swift
+var names = [ "Sean", "Laura", "Dave", "Crusty" ]
+names.sort()
+print(names[3])
+```
+
+In the example above, how do we know that last line is OK? Our thought
+process is something like this:
+
+- We started with values in indices zero through three of `names`.
+- sorting `names` rearranges elements without changing the set of valid indices.
+- Therefore, in the last line, `names` still has an element at index `3`.
+
+That's just an informal proof, and we use that kind of reasoning every
+time we write a line of code. So, regular programming is on the same
+continuum as formally proving correctness.  We're going to explore the
+power that's available by moving toward the formal end of that
+spectrum to see how it enables Better Code.
+
+## Local Reasoning
+
+The kind of everyday thinking demonstrated above is only practical if
+we can reason locally about code.  The clearest definition of Local
+Reasoning we've found is [this
+one](https://medium.com/@nathangitter/local-reasoning-in-swift-6782e459d)
+from Nathan Gitter: “Local reasoning is the idea that the reader can
+make sense of the code directly in front of them, without going on a
+journey discovering how the code works.”
+
+So in our example, what we know about `sort` allow us to reason about
+its use without looking at its implementation.  In fact, local
+reasoning is so fundamental that most of our programming best
+practices exist just to support our ability to do it.  It's why we
+use access control, why we break programs into components like
+functions, types, and modules, and why we try to keep them small.
+
+## Components and Abstraction
+
+Obviously, “knowing something about `sort`” is only possible and
+relevant because `sort` is a separate component with its own name. If
+we had injected the sorting implementation directly into our code,
+we'd have to reason about all of its details, and we'd need a comment
+to tell us that the implementation was sorting the elements.
```

Comment:
> we'd need a comment to tell us that the implementation was sorting the elements.

Would we? That's probably true if you've already accepted the DbC ethos, but at this point in the narrative, I'm not sure we have. If we've already "reason[ed] about all of its details", would we actually need a comment to tell us what it's doing? IIUC we only need a comment to know if its behavior is correct.

Reply:
Sorry, I don't mean to be pedantic, but even if sorting is complicated, you've already stated we've "reason[ed] about all of its details". What does it mean to reason about all details if not to understand a (yes, complicated) piece of code? In practice I hear you that a comment would be _helpful_, but as the sentence is structured I hear a contradiction. Does that make sense?